docs/adr/0006_component_constructors.md

```diff
@@ -0,0 +1,334 @@
+    GetTargetRepository(ctx context.Context, *constructor.Component) (TargetRepository, error)
+}
+
+type ResourceRepository interface {
+    // DownloadResource downloads a resource from the repository.
+    DownloadResource(ctx context.Context, res *descriptor.Resource) (content blob.ReadOnlyBlob, err error)
+}
+
+type ResourceRepositoryProvider interface {
+    // ResourceRepositoryProvider returns the target ocm resource repository for the given component specification in the constructor.
+    GetResourceRepository(ctx context.Context, *constructor.Component) (Repository, error)
+}
+
+// Options are the options for construction based on a *constructor.Constructor.
+type Options struct {
+    // While constructing a component version, the constructor library will use the given target repository provider
+    // to get the target repository for the component specification.
+    TargetRepositoryProvider
+    
+    // While constructing a component version, the constructor library will use the given resource repository provider
+    // to get the resource repository for the component specification when processing resources by value.
+    ResourceRepositoryProvider
+    
+    // While constructing a component version, the constructor library will use the given resource input method provider
+    // to get the resource input method for the component specification when processing resources with an input method.
+    ResourceInputMethodProvider
+    
+    // While constructing a component version, the constructor library will use the given source input method provider
+    // to get the source input method for the component specification when processing sources with an input method.
+    SourceInputMethodProvider
+
+	// While constructing a component version, the constructor library will use the 
+	// given function to decide whether a resource should be processed by value or not.
+    ProcessByValue func(*constructor.Resource) bool
+}
+
+// Construct constructs a component version based on the given specification and options.
+func Construct(ctx context.Context, specification *constructor.Constructor, opts Options) ([]*descriptor.Descriptor, error)
+```
+
+## Pros and Cons
+
+Pros:
+
+* Very modular and encapsulated constructor design
+* Extensible and easy to add new input methods by adjusting the providers
+* Very easy control flow that can be made concurrent
+
+Cons:
+
+* Complex Interfaces for a lot of different scenarios (input and access specs are mutually exclusive and lead to branching)
+
+## Conclusion
+
+The new component constructor library will:
+
+- Provide a simple and extensible way to construct component versions based on a `component-constructor.yaml` specification.
+- Support various input methods and access specifications, and upload the resulting component version to an OCM repository.
+- Download resources and upload them as local blobs.
+- Process component version resources and sources in parallel to improve efficiency.
```

Comment:
You mean that it will download things in parallel right? Because for the construction I think it needs to be sequential.

Reply:
for construction, resources and source inputs can be processed concurrently I believe. What makes you think otherwise?

Reply:
I have a distinct memory, that the construction of the component descriptor needed to follow a strict order. But maybe that memory is flawed. :)

Reply:
Are you referring to the processing of individual resources and their order within a component version? Im not aware of such logic, but im happy to accomodate anything in the ADR that you know of or can point me to!

Reply:
Meh, it's fine. I was looking for it, but couldn't find anything that would say we can't do something like that. So this is fine.